File: lab3_banker_algorithm/src/python/main.py
import json
import logging
from pathlib import Path

from banker import Banker


def load_json(json_path):
    with open(json_path, "r") as f:
        return json.load(f)


def main():
    # Set input and output file paths.
    BASE_DIR = Path(__file__).resolve().parent
    DATA_DIR = BASE_DIR.parent.parent / "data"
    JSON_FILE = DATA_DIR / "example_1.json"
    LOG_FILE = DATA_DIR / "output.log"

    # Configure the logging format.
    logging.basicConfig(
        level=logging.INFO,
        # format="[%(asctime)s] %(message)s",
        format="%(message)s",
        handlers=[
            logging.FileHandler(LOG_FILE, mode="w", encoding="utf-8"),
            logging.StreamHandler(),
        ],
    )

    logging.info(f"Read data from file {JSON_FILE}.")
    data = load_json(JSON_FILE)

    banker = Banker(data["total"], data["allocation"], data["max_demand"])
    banker.print_state()  # Print the initial state.

    for req in data.get("requests", []):
        pid = req["pid"]
        request = req["request"]
        banker.process_request(pid, request)
        banker.print_state()


if __name__ == "__main__":
    main()

File: lab3_banker_algorithm/src/main.cpp
#include <iostream>
#include <string>
#include <vector>

#include "../include/banker.hpp"
#include "../include/utils.hpp"

int main(int argc, char *argv[])
{
    std::string json_path = (argc > 1) ? argv[1] : "../data/example_1.json";

    std::vector<int> total;
    std::vector<std::vector<int>> allocation;
    std::vector<std::vector<int>> max_demand;
    std::vector<Request> requests;

    if (!utils::load_data_from_json(json_path, total, allocation, max_demand, requests))
    {
        return 1;
    }

    Banker banker(total, allocation, max_demand);
    banker.print_state();

    for (const auto &req : requests)
    {
        banker.process_request(req.pid, req.request);
        banker.print_state();
    }

    return 0;
}